src/Juvix/Compiler/Internal/Translation/FromConcrete.hs

```diff
@@ -764,6 +765,74 @@ goExpression = \case
+            updateKind :: Internal.Subs =
+              HashMap.fromList
+                [ (s, Internal.toExpression s') | s <- args, let s' = Internal.IdenFunction (set Internal.nameKind KNameFunction s)
+                ]
+            napp' =
+              Concrete.NamedApplication
+                { _namedAppName = napp ^. namedApplicationNewName,
+                  _namedAppArgs = nonEmpty' $ createArgumentBlocks (sig ^. nameSignatureArgs)
+                }
+        e <- goNamedApplication napp'
+        let expr =
+              Internal.substitutionE updateKind
+                . Internal.ExpressionLet
+                $ Internal.Let
+                  { _letClauses = cls,
+                    _letExpression = e
+                  }
+        Internal.clone expr
+        where
+          goArgs :: NonEmpty (NamedArgumentNew 'Scoped) -> Sem r (NonEmpty Internal.LetClause)
+          goArgs args = nonEmpty' . mkLetClauses <$> mapM goArg args
+            where
+              goArg :: NamedArgumentNew 'Scoped -> Sem r Internal.PreLetStatement
+              goArg = fmap Internal.PreLetFunctionDef . goFunctionDef . (^. namedArgumentNewFunDef)
+
+          -- TODO: error on unprocessed args
+          createArgumentBlocks :: [NameBlock 'Scoped] -> [ArgumentBlock 'Scoped]
+          createArgumentBlocks sblocks = snd $ foldr goBlock (args0, []) sblocks
+            where
+              args0 :: HashSet S.Symbol = HashSet.fromList $ fmap (^. namedArgumentNewFunDef . signName) (toList appargs)
+              goBlock :: NameBlock 'Scoped -> (HashSet S.Symbol, [ArgumentBlock 'Scoped]) -> (HashSet S.Symbol, [ArgumentBlock 'Scoped])
+              goBlock NameBlock {..} (args, blocks)
+                | null names = (args', blocks)
+                | otherwise = (args', block' : blocks)
+                where
+                  names =
+                    HashSet.intersection
+                      (HashSet.fromList $ HashMap.keys _nameBlock)
+                      (HashSet.map (^. S.nameConcrete) args)
+                  args' = HashSet.filter (not . flip HashSet.member names . (^. S.nameConcrete)) args
+                  nargs = nonEmpty' $ map goArg (toList names)
+                  block' =
+                    ArgumentBlock
+                      { _argBlockDelims = Irrelevant Nothing,
+                        _argBlockImplicit = _nameImplicit,
+                        _argBlockArgs = nargs
+                      }
+                  goArg :: Symbol -> NamedArgument 'Scoped
+                  goArg sym =
+                    NamedArgument
+                      { _namedArgName = sym,
+                        _namedArgAssignKw = Irrelevant dummyKw,
+                        _namedArgValue = Concrete.ExpressionIdentifier $ ScopedIden name Nothing
+                      }
+                    where
+                      name = over S.nameConcrete NameUnqualified $ fromJust $ find ((== sym) . (^. S.nameConcrete)) args
```

Comment:
As `sym` ultimately comes from `args` itself, could `sym` and `name` be associated somehow during processing so it doesn't need to be obtained using `fromJust . find` at the end?

Reply:
I created a HashMap associating symbols with names